src/serac/physics/utilities/functional/domain_integral_cuda.cuh

```diff
@@ -0,0 +1,292 @@
+#pragma once
+
+#include "mfem.hpp"
+#include "mfem/linalg/dtensor.hpp"
+
+#include "serac/physics/utilities/functional/integral_utilities.hpp"
+#include "serac/physics/utilities/functional/domain_integral_shared.hpp"
+#include <cstring>
+
+namespace serac {
+
+namespace detail {
+
+/**
+ * @brief utility method to display last cuda error message
+ *
+ * @param[in] o The output stream to post success or CUDA error messages
+ * @param[in] success_string A string to print if there are no CUDA error messages
+ */
+inline void displayLastCUDAErrorMessage(std::ostream& o, const char* success_string = "")
+{
+  auto error = cudaGetLastError();
+  if (error != cudaError::cudaSuccess) {
+    o << "Last CUDA Error Message :" << cudaGetErrorString(error) << std::endl;
+  } else if (strlen(success_string) > 0) {
+    o << success_string << std::endl;
+  }
+}
+
+/**
+ * @brief Defines whether to loop by elements or quadrature points
+ */
+enum ThreadExecutionPolicy
+{
+  THREAD_PER_ELEMENT_QUADRATURE_POINT,
+  THREAD_PER_ELEMENT
+};
+
+/**
+ * @brief Contains the GPU launch configuration
+ */
+struct ThreadExecutionConfiguration {
+  int blocksize;
+};
+
+}  // namespace detail
+
+namespace domain_integral {
+
+template <Geometry g, typename test, typename trial, int Q, typename derivatives_type, typename lambda, typename u_type,
+          typename r_type, typename J_type, typename X_type>
+__global__ void eval_cuda_element(const u_type u, r_type r, derivatives_type* derivatives_ptr, J_type J, X_type X,
+                                  int num_elements, lambda qf)
+{
+  using test_element          = finite_element<g, test>;
+  using trial_element         = finite_element<g, trial>;
+  using element_residual_type = typename trial_element::residual_type;
+  static constexpr auto rule  = GaussQuadratureRule<g, Q>();
+
+  // for each element in the domain
+  const int grid_stride = blockDim.x * gridDim.x;
+#pragma unroll
+  for (int e = blockIdx.x * blockDim.x + threadIdx.x; e < num_elements; e += grid_stride) {
+    // get the DOF values for this particular element
+    auto u_elem = detail::Load<trial_element>(u, e);
+
+    // this is where we will accumulate the element residual tensor
+    element_residual_type r_elem{};
+
+    // for each quadrature point in the element
+    for (int q = 0; q < static_cast<int>(rule.size()); q++) {
+      eval_quadrature<g, test, trial, Q, derivatives_type, lambda>(e, q, u_elem, r_elem, derivatives_ptr, J, X,
+                                                                   num_elements, qf);
+    }
+
+    // once we've finished the element integration loop, write our element residuals
+    // out to memory, to be later assembled into global residuals by mfem
+    detail::Add(r, r_elem, e);
+  }  // e loop
+}
+
+template <Geometry g, typename test, typename trial, int Q, typename derivatives_type, typename lambda, typename u_type,
```

Comment:
While not much, this duplication is a bit worrisome. I assume this will be raja-ified and merged with the CPU version in a future PR.

Reply:
Sorry, I don't think I explained it well, since this is an intermediate work PR. I'll remember to add in some comments here. `eval_cuda_element` and `eval_cuda_quadrature` are GPU kernels that launch over 1 thread per element and 1 thread per quadrature point. The kernels will be slightly different and you may want to use one over another depending on the actual lambda `qf`. `eval_quadrature` is the same for both the CPU and the GPU. Basically in the `eval_cuda_quadrature` the loop over quadrature points is flattened and merged into the loop over elements. When you do that, you need to retrieve `e` and `q`. It may be possible to merge them into RAJA, but it might be a little harder to control the performance of the kernels.

`eval_cuda_kernel` is the actual call that launches the kernel on the GPU and converts `mfem::Vectors/DenseTensors` to something the kernel can understand.